### PR TITLE
[SR-3811] optimize high card count distinct

### PR DESCRIPTION
**Pull request by dirtysalt**
Use another way to serialize/deserialize hahset in distinct agg state:
1. save packet size
2. better cache locality to do merge

Testing following SQL on ssb dataset, query time can be reduced from 60s -> 20s

```
SELECT COUNT (DISTINCT lo_orderkey), lo_linenumber FROM lineorder_flat group by lo_linenumber;
```

### CODE REVIEW

**Comment by imay** [comment]
@dirtysalt 
I have one question. If this CL has been merged, can it be compatible with old version.
If not this will make our program can not grayscale upgrade

**Comment by dirtysalt** [comment]
> @dirtysalt
> I have one question. If this CL has been merged, can it be compatible with old version.
> If not this will make our program can not grayscale upgrade

Unfortunately not yet. But I think we can do something hacky to make it compatible. 

Right serialize dump format is like,

```
| size | capacity | control array | slot array|
```

and size is `size_t`, in most case it should be less than `(1<<63)-1`(otherwise this hashtable is huge). So we can tell this is old serialized version if `size` msb is 0. And for further version, we put set msb of `size` to 1.

**Comment by dirtysalt** [comment]
> @dirtysalt
> I have one question. If this CL has been merged, can it be compatible with old version.
> If not this will make our program can not grayscale upgrade

yeah, it's very hard to be backward compatible. Because old version of BE does not recognize new data format. So basically yes, it's can not be upgraded in a grayscale way.

**Comment by imay** [comment]
> > @dirtysalt
> > I have one question. If this CL has been merged, can it be compatible with old version.
> > If not this will make our program can not grayscale upgrade
> 
> yeah, it's very hard to be backward compatible. Because old version of BE does not recognize new data format. So basically yes, it's can not be upgraded in a grayscale way.

If we can not upgrade in a grayscale way, this would be bad for users.
So we should make it back compatible.

File: be/src/exprs/agg/distinct.h [hunk 1]
```diff
@@ -53,26 +53,33 @@ struct DistinctAggregateState<PT, FixedLengthPTGuard<PT>> {
...
+        size_t size = 0;
+        memcpy(&size, src, sizeof(size));
+        set.rehash(set.size() + size);
+
+        src += sizeof(size);
+        const T* data = reinterpret_cast<const T*>(src);
+        static const size_t prefetch_dist = 4;
+        for (size_t i = 0; i < size; i++) {
+            if ((i + prefetch_dist) < size) {
+                set.prefetch(data[i]);
```
Comment (DeepThinker666) [inline be/src/exprs/agg/distinct.h:78]: data[i + prefetch_dist]?

**Comment by DeepThinker666** [comment]
could you add some background for this pr?

File: be/src/exprs/agg/distinct.h (hunk 1 above)
Comment (dirtysalt) [reply be/src/exprs/agg/distinct.h:78]: right, thanks.

**Comment by dirtysalt** [comment]
> could you add some background for this pr?

Right, we are doing some benchmark comparisons with CK. And we find out that we don't have good performance when we are doing (count/distinct + group by) those cases. There are several optimization points:
1. better hashtable for inserting and finding
2. serialization and deserialization.

File: be/src/column/hash_set.h [hunk 2]
```diff
@@ -35,6 +35,32 @@ class EqualOnSliceWithHash {
     bool operator()(const SliceWithHash& x, const SliceWithHash& y) const {
         // by comparing hash value first, we can avoid comparing real data
         // which may touch another memory area and has bad cache locality.
+        if (x.size != y.size) return false;
```
Comment (stdpain) [inline be/src/column/hash_set.h:38]:
```suggestion
        return x.size != y.size && x.hash == y.hash && memequal(x.data, x.size, y.data, y.size);
```

Comment (stdpain) [reply be/src/column/hash_set.h:38]: I'm not so sure this will reduce some of the control dependency

Comment (Seaven) [reply be/src/column/hash_set.h:38]: It's useless, use && can't reduce control dependency

Comment (stdpain) [reply be/src/column/hash_set.h]: It seems`&&` can't reduce control dependency, but it seem it wil generate shorter assemble ... 
https://godbolt.org/z/9rE5nosqo

Comment (dirtysalt) [reply be/src/column/hash_set.h]: Those 3 versions are almost the same 17 lines.  I'm not sure which will perform better in production. 

And I guess this is assembly code generated in an independent function.  In production, those functions will be inlined and I'm not sure the final assembly code in production.

Anyway,  I'm okay to use a compact form like `x.size == y.size && x.hash == y.hash`.

**Comment by dirtysalt** [comment]
running benchmark on ssb_100g with max_threads = 16. 

| Query    | CK    | SR Master | SR PR | SR/CK |
|----------|-------|-----------|-------|-------|
| Q01      | 16739 | 1110      | 1086  | 0.065 |
| Q02      | 738   | 1052      | 1069  | 1.449 |
| Q03      | 482   | 400       | 360   | 0.747 |
| Q04      | 175   | 176       | 174   | 0.994 |
| Q05      | 52141 | 15840     | 9526  | 0.183 |
| Q06      | 1682  | 2305      | 2037  | 1.211 |
| Q07      | 966   | 1347      | 1261  | 1.305 |
| Q08      | 351   | 326       | 337   | 0.96  |
| Q09      | 51107 | 6277      | 4898  | 0.096 |
| Q10      | 1920  | 2449      | 2189  | 1.14  |
| Q11      | 33665 | 6011      | 4749  | 0.141 |
| Q12      | 1728  | 1958      | 1832  | 1.06  |
| Q13      | 32919 | 10079     | 9475  | 0.288 |
| Q14      | 5290  | 6816      | 6764  | 1.279 |
| Q15      | 21193 | 9766      | 9243  | 0.436 |
| Q16      | 4976  | 6003      | 5844  | 1.174 |

```
# 100G
# lo_orderkey = 150,000,000
# lo_partkey = 1000,000
# lo_suppkey = 200,000
# lo_quantity = 50

# lo_linenumber = 7
# lo_supplycost = 15933
# lo_linenumber, lo_supplycost = 111531

-- Q01
select count (distinct lo_orderkey) from lineorder_flat;
-- Q02
SELECT COUNT (DISTINCT lo_partkey) FROM lineorder_flat;
-- Q03
SELECT COUNT (DISTINCT lo_suppkey) FROM lineorder_flat;
-- Q04
SELECT COUNT (DISTINCT lo_quantity) FROM lineorder_flat;

-- Q05
SELECT COUNT (DISTINCT lo_orderkey), lo_linenumber FROM lineorder_flat group by lo_linenumber;
-- Q06
SELECT COUNT (DISTINCT lo_partkey), lo_linenumber FROM lineorder_flat group by lo_linenumber;
-- Q07
SELECT COUNT (DISTINCT lo_suppkey), lo_linenumber FROM lineorder_flat group by lo_linenumber;
-- Q08
SELECT COUNT (DISTINCT lo_quantity), lo_linenumber FROM lineorder_flat group by lo_linenumber;

-- Q09
SELECT COUNT (DISTINCT lo_orderkey), lo_supplycost FROM lineorder_flat group by lo_supplycost;
-- Q10
SELECT COUNT (DISTINCT lo_partkey), lo_supplycost FROM lineorder_flat group by lo_supplycost;
-- Q11
SELECT COUNT (DISTINCT lo_suppkey), lo_supplycost FROM lineorder_flat group by lo_supplycost;
-- Q12
SELECT COUNT (DISTINCT lo_quantity), lo_supplycost FROM lineorder_flat group by lo_supplycost;

-- Q13
SELECT COUNT (DISTINCT lo_orderkey), lo_linenumber, lo_supplycost FROM lineorder_flat group by lo_linenumber, lo_supplycost;
-- Q14
SELECT COUNT (DISTINCT lo_partkey), lo_linenumber, lo_supplycost FROM lineorder_flat group by lo_linenumber, lo_supplycost;
-- Q15
SELECT COUNT (DISTINCT lo_suppkey), lo_linenumber, lo_supplycost FROM lineorder_flat group by lo_linenumber, lo_supplycost;
-- Q16
SELECT COUNT (DISTINCT lo_quantity), lo_linenumber, lo_supplycost FROM lineorder_flat group by lo_linenumber, lo_supplycost;
```

File: gensrc/thrift/PlanNodes.thrift [hunk 3]
```diff
@@ -469,6 +469,8 @@ struct TAggregationNode {
   // For profile attributes' printing: `Grouping Keys` `Aggregate Functions`
   22: optional string sql_grouping_keys
   23: optional string sql_aggregate_functions
+
+  24: optional i32 agg_func_set_version = 1
```
Comment (kangkaisen) [inline gensrc/thrift/PlanNodes.thrift:473]: disticnt_function_version?

File: be/src/column/hash_set.h [hunk 4]
```diff
@@ -35,7 +35,31 @@ class EqualOnSliceWithHash {
     bool operator()(const SliceWithHash& x, const SliceWithHash& y) const {
         // by comparing hash value first, we can avoid comparing real data
         // which may touch another memory area and has bad cache locality.
-        return x.hash == y.hash && memequal(x.data, x.size, y.data, y.size);
+        return x.size == y.size && x.hash == y.hash && memequal(x.data, x.size, y.data, y.size);
```
Comment (kangkaisen) [inline be/src/column/hash_set.h:38]: memequal will check `x.size == y.size` firstly.

File: be/src/column/hash_set.h [hunk 5]
```diff
@@ -35,7 +35,31 @@ class EqualOnSliceWithHash {
...
+    std::size_t operator()(const TSliceWithHash<seed>& slice) const { return slice.hash; }
+};
+
+template <PhmapSeed seed>
+class TEqualOnSliceWithHash {
+public:
+    bool operator()(const TSliceWithHash<seed>& x, const TSliceWithHash<seed>& y) const {
+        // by comparing hash value first, we can avoid comparing real data
+        // which may touch another memory area and has bad cache locality.
+        return x.size == y.size && x.hash == y.hash && memequal(x.data, x.size, y.data, y.size);
```
Comment (kangkaisen) [inline be/src/column/hash_set.h:62]: memequal will check x.size == y.size firstly.

**Review by trueeyu — APPROVED** [review]

**Review by satanson — APPROVED** [review]

File: be/src/column/hash_set.h (hunk 2 above)
Comment (Seaven) [reply be/src/column/hash_set.h]: You can try `x.size == y.size & x.hash == y.hash & memequal(x.data, x.size, y.data, y.size);`,  `&` can reduce control dependency,  more friendly to cache, but there is no short circuit, and I not sure will faster than now.

**Comment by SeaAndHillMe** [comment]
good to learn

**Pull request merged**